File: projects/intersect/intersect.cpp
#include <boost/python.hpp>
#include "boost/numpy.hpp"
#include <boost/python/numeric.hpp>
#include <boost/python/suite/indexing/vector_indexing_suite.hpp>
#include <vector>
#include <string>
#include <iostream>
#include "common/common.h"
#include "detector/detector.h"
#include <fstream>

#include <boost/archive/text_oarchive.hpp>
#include <boost/archive/text_iarchive.hpp>
#include <boost/archive/binary_oarchive.hpp>
#include <boost/archive/binary_iarchive.hpp>
#include <boost/serialization/vector.hpp>

using namespace std;
using namespace boost::python;

using intersect::Keypoint;
using intersect::Image;

using intersect::Detector;

BOOST_PYTHON_MODULE(intersect)
{
    //using namespace boost::python;
    //boost::python::numeric::array::set_module_and_type("numpy", "ndarray");

    boost::numpy::initialize();  // have to put this in any module that uses Boost.NumPy

    // class Keypoint
    class_<Keypoint>("Keypoint", init<float, float, float>())
        .def_readwrite("x", &Keypoint::x)
        .def_readwrite("y", &Keypoint::y)
        .def_readwrite("scale", &Keypoint::scale)
        .def("__repr__", &Keypoint::toString)
        ;

    // class Image
    class_<Image>("Image", init<int, int, const boost::numpy::ndarray&>())
        .def_readwrite("width", &Image::width)
        .def_readwrite("height", &Image::height)
        .def_readwrite("data", &Image::data)
        .def("fuck", &Image::fuck)
        .def("getData", &Image::getCharData)
        .def("getBuffer", &Image::getBuffer)
        .def("saveToRedis", &Image::saveToRedis)
        ;

    class_<std::vector<Keypoint> >("KeypointVector")
        .def(vector_indexing_suite<std::vector<Keypoint> >())
        ;

    // class Detector
    class_<Detector>("Detector")
        .def("detect", &Detector::detect)
        .def("fake_detect", &Detector::fakeDetect)
        .def("loadKeypoints", &Detector::loadKeypoints)
        .def("saveKeypoints", &Detector::saveKeypoints)
        .def_readwrite("keypoints", &Detector::keypoints)
        ;
}

File: projects/intersect/intersect.py
#!/usr/bin/env python
import intersect
from intersect import Keypoint, Image, Detector, Extractor
import numpy as np

#kps = [Keypoint(1.0, 2.0, 3.0),
        #Keypoint(4.0, 5.0, 6.0),
        #Keypoint(7.0, 8.0, 9.0)]

# construct image
data = np.array([10, 20, 30, 40, 50, 60], dtype=np.uint8)
image = Image(3, 2, data)
image.fuck()

#image = Image(3, 2)
# keypoints[2] = Keypoint(1000,2000,3000)

# detect keypoints
detector = Detector()

detector.detect(image)
keypoints = detector.keypoints
    
filename = "keypoints.bin"
keypoints = detector.keypoints

# get features from keypoints
extractor = Extractor()
extractor.extract(keypoints)

features = extractor.getFeatures()

extractor.fuckWithFeatures()

print ""

print "Python:"
print "------------------------------------"
print features

#for kp in keypoints:
    #print kp

